### PR TITLE
Enable preventing engine build at runtime

### PR DESCRIPTION
**Pull request by pooyadavoodi**
Adds a new API argument `allow_build_at_runtime` which allows users to prevent building TRT engines at runtime if desired. The existing behavior can be achieved by setting this argument to `True` which is also the default value.

This argument is useful for users of the `build()` method that try to build all the engines offline before doing any inference, and then want to avoid any optimization during inference, keeping low latency. In such cases, if there is no engine found in the cache, then native TF is used instead.

### CODE REVIEW

**Review by toom5556 — DISMISSED** [review]

**Comment by pooyadavoodi** [comment]
I just fixed the api compatibility test by adding the new argument to the golden api.

File: tensorflow/python/compiler/tensorrt/trt_convert.py [hunk 1]
```diff
@@ -1098,6 +1119,25 @@ def _serialize_and_track_engine(node):
     signatures = {
         key: value for key, value in self._saved_model.signatures.items()
     }
+
+    # Set allow_build_at_runtime=False if asked by user.
+    # This attribute is set here because build() needs it to be True
+    # in order to build engines.
+    if not self._conversion_params.allow_build_at_runtime:
+      def _reset_allow_build_at_runtime(node):
+        node.attr["allow_build_at_runtime"].b = False
```
Comment (pooyadavoodi) [inline tensorflow/python/compiler/tensorrt/trt_convert.py]: @aaroey shouldn't this be `_allow_build_at_runtime`?
I wonder why tests don't fail.

**Comment by aaroey** [comment]
Hi @pooyadavoodi, could you help to resolved the conflicts? Thanks.

**Comment by pooyadavoodi** [comment]
This PR has to wait until https://github.com/tensorflow/tensorflow/pull/35198 gets merged.

**Comment by pooyadavoodi** [comment]
@sanjoy @aaroey 

When I run any c++ test that executes native TF function instead of TRT, then I get this error: `Executor failed to create kernel. Not found: No registered '_Arg' OpKernel for 'GPU' devices compatible with node`.

I don't think it's related to `_Arg` because I replaced it with other nodes and still fails.

It also seems unrelated to this PR because I can repro the error without the PR if I change the test to force running native TF functions. For example, you can substitute `ops::Add` by `ops::Less` in https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/tf2tensorrt/kernels/trt_engine_op_test.cc#L60 to repro the issue. We currently don't have any tests that execute native TF functions and that's why we never caught this.

This is how I run the test: `bazel test -c opt --config=cuda --cxxopt=-D_GLIBCXX_USE_CXX11_ABI=0 --noincompatible_do_not_split_linking_cmdline //tensorflow/compiler/tf2tensorrt:trt_engine_op_test`
Let me know if the problem is how I run the test.

I am kind of stuck in root causing this.

**Comment by sanjoy** [comment]
@pooyadavoodi It could be that you're not linking in the correct kernel registrations.  For instance, the GPU implementation of the `_Arg` op is [defined](https://github.com/tensorflow/tensorflow/blob/45e1e4598d3ebab316bf87df9160656f524af36c/tensorflow/core/kernels/function_ops.cc#L105) in function_ops.cc and you need to link that in to be able to execute `_Arg` ops.

Can you try linking in `//tensorflow/core/kernels:function_ops` and `//tensorflow/core/kernels:array` to the test binary and see if that fixes or at least changes the failure?

**Comment by pooyadavoodi** [comment]
> @pooyadavoodi It could be that you're not linking in the correct kernel registrations. For instance, the GPU implementation of the `_Arg` op is [defined](https://github.com/tensorflow/tensorflow/blob/45e1e4598d3ebab316bf87df9160656f524af36c/tensorflow/core/kernels/function_ops.cc#L105) in function_ops.cc and you need to link that in to be able to execute `_Arg` ops.
> 
> Can you try linking in `//tensorflow/core/kernels:function_ops` and `//tensorflow/core/kernels:array` to the test binary and see if that fixes or at least changes the failure?

Thanks @sanjoy. Linking changed the error. Now I am looking at another problem.

**Comment by sanjoy** [comment]
> Thanks @sanjoy. Linking changed the error. Now I am looking at another problem.

Looks like this isn't ready for review so removing the tag.  @pooyadavoodi please LMK if this is indeed ready for review.

**Comment by pooyadavoodi** [comment]
@sanjoy I am seeing the following error when running `trt_engine_op_test`:

```
unknown file: Failure
C++ exception with description "bad_function_call" thrown in the test body.
```
The error happens when TF-TRT decides to execute the native TF function instead of the TRT engine. Looks like the TF function is not available?

We register the graph in the function library here: https://github.com/pooyadavoodi/tensorflow/blob/allow_build_at_runtime/tensorflow/compiler/tf2tensorrt/kernels/trt_engine_op_test.cc#L69-L71

Any idea what could be the reason?

**Comment by sanjoy** [comment]
https://en.cppreference.com/w/cpp/utility/functional/bad_function_call says "`std::bad_function_call` is the type of the exception thrown by `std::function::operator()` if the function wrapper has no target.".  So probably we have an uninitialized `std::function` somewhere.  Can you get the stack trace from where the exception is thrown?

**Comment by pooyadavoodi** [comment]
> https://en.cppreference.com/w/cpp/utility/functional/bad_function_call says "`std::bad_function_call` is the type of the exception thrown by `std::function::operator()` if the function wrapper has no target.". So probably we have an uninitialized `std::function` somewhere. Can you get the stack trace from where the exception is thrown?

`bazel test` is not showing the stack trace. Do you know how I can force it to show the stack trace?

**Comment by sanjoy** [comment]
> `bazel test` is not showing the stack trace. Do you know how I can force it to show the stack trace?

You could try running the program under GDB and put a breakpoint on the constructor of `std::bad_function_call` to see where it is constructed.

**Comment by bixia1** [comment]
I still see the "bad_function_call" error when running the test that you added in the PR, here is the stack:
*** SIGABRT received by PID 2361 (TID 2361) from PID 2361; stack trace: ***
PC: @     0x7f1fd1132602  (unknown)  raise
    @     0x55b930fbd67b       1728  FailureSignalHandler()
    @     0x7f1fd12b09a0  1950672896  (unknown)
    @     0x55b92b3e2109         16  std::__u::__throw_bad_function_call()
    @     0x55b92f737cc9         16  std::__u::__function::__policy_invoker<>::__call_empty()
    @     0x55b92f730c0b        512  tensorflow::(anonymous namespace)::ExecutorImpl::RunAsync()
    @     0x55b92f74452d        560  tensorflow::FunctionLibraryRuntimeImpl::Run()
    @     0x55b92b3eccf3        336  tensorflow::tensorrt::TRTEngineOp::ExecuteNativeSegment()
    @     0x55b92b3ee2cd        192  tensorflow::tensorrt::TRTEngineOp::ComputeAsync()
    @     0x55b92fb4703d         80  tensorflow::AsyncOpKernel::Compute()
    @     0x55b92f6bab00        400  tensorflow::BaseGPUDevice::Compute()
    @     0x55b92b3e0814        208  tensorflow::OpsTestBase::RunOpKernel()
    @     0x55b92b3e09d8        160  tensorflow::tensorrt::TRTEngineOpTestBase_AllowBuildAtRuntime_Test::TestBody()
    @     0x55b92cda00c2         48  testing::Test::Run()
    @     0x55b92cda0d8c         96  testing::TestInfo::Run()
    @     0x55b92cda16c7         80  testing::TestSuite::Run()
    @     0x55b92cdaeb47        208  testing::internal::UnitTestImpl::RunAllTests()
    @     0x55b92cdae16f         64  testing::UnitTest::Run()
    @     0x55b92ca08f6e         32  main
    @     0x7f1fd111ebbd        208  __libc_start_main

**Comment by pooyadavoodi** [comment]
I managed to make a debug build of TF that shows stack trace when it crashes but still doesn't show any backtrace in gdb. It would be good to document how to make a debug build of TF.

Here is the full log of the test: https://gist.github.com/pooyadavoodi/3addaeff4cca5f2de6ecfa8d8fdd00bd

My guess is that the failure happens because of `"Not found: No attr named` such as:
```
2020-01-31 18:36:40.777739: I tensorflow/core/platform/status.cc:95] Generated non-OK status: "Not found: No attr named '_scoped_allocator' in NodeDef:". *** Begin stack trace ***
	tensorflow::CurrentStackTrace()
	tensorflow::Status::Status(tensorflow::error::Code, absl::string_view)
	tensorflow::AttrSlice::Find(absl::string_view, tensorflow::AttrValue const**) const
	tensorflow::GetNodeAttr(tensorflow::AttrSlice const&, absl::string_view, std::vector<int, std::allocator<int> >*)
	
	tensorflow::NewLocalExecutor(tensorflow::LocalExecutorParams const&, tensorflow::Graph const&, tensorflow::Executor**)
	
	tensorflow::NewExecutor(std::string const&, tensorflow::LocalExecutorParams const&, tensorflow::Graph const&, std::unique_ptr<tensorflow::Executor, std::default_delete<tensorflow::Executor> >*)
	tensorflow::FunctionLibraryRuntimeImpl::CreateItem(tensorflow::FunctionLibraryRuntimeImpl::Item**)
	tensorflow::FunctionLibraryRuntimeImpl::GetOrCreateItem(unsigned long long, tensorflow::FunctionLibraryRuntimeImpl::Item**)
	tensorflow::FunctionLibraryRuntimeImpl::Run(tensorflow::FunctionLibraryRuntime::Options const&, unsigned long long, absl::Span<tensorflow::Tensor const>, std::vector<tensorflow::Tensor, std::allocator<tensorflow::Tensor> >*, std::function<void (tensorflow::Status const&)>)
	tensorflow::tensorrt::TRTEngineOp::ExecuteNativeSegment(tensorflow::OpKernelContext*, tensorflow::tensorrt::AsyncHelper*)
	tensorflow::tensorrt::TRTEngineOp::ComputeAsync(tensorflow::OpKernelContext*, std::function<void ()>)
	tensorflow::AsyncOpKernel::Compute(tensorflow::OpKernelContext*)
	tensorflow::BaseGPUDevice::Compute(tensorflow::OpKernel*, tensorflow::OpKernelContext*)
	
	void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*)
	testing::Test::Run()
	testing::TestInfo::Run()
	testing::TestSuite::Run()
	testing::internal::UnitTestImpl::RunAllTests()
	bool testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*)
	testing::UnitTest::Run()
	main
	__libc_start_main
	
*** End stack trace ***
```

**Comment by pooyadavoodi** [comment]
We know that TF-TRT source code can successfully execute native segments (TF functions added to graph as a function library). The problem is that our c++ tests can'r run the native segments.

TF-TRT Source code: 
- Functions are added to the graph here:  https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/tf2tensorrt/convert/convert_graph.cc#L672
- Function libraries are instantiated here: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/tf2tensorrt/kernels/trt_engine_op.cc#L279
 
TF-TRT c++ unit tests (e.g. trt_engine_op_test):
- Functions are added to the graph here:  https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/tf2tensorrt/kernels/trt_engine_op_test.cc#L69
- Function libraries are not instantiated in the same place as TF-TRT source code because the test creates a new instance of TRTEngineOp which triggers its constructor which is where the function libraries get instantiated.

**Comment by sanjoy** [comment]
I triaged the problem down to `OpsTestBase`.  With this [change](https://gist.github.com/sanjoy/2f355116c7ca88bfcd8665ba988a1bbf) patched in I now get a different [failure](https://gist.github.com/sanjoy/ba102ae122d2eaaa2073e5afe0ee0de2) from the test.

**Comment by cliffwoolley** [comment]
@nluehr @cliffwoolley

**Comment by pooyadavoodi** [comment]
@sanjoy I can see the following error in the log you linked:

```
2020-02-01 06:22:10.737479: E tensorflow/compiler/tf2tensorrt/utils/trt_logger.cc:42] DefaultLogger ../rtSafe/safeContext.cpp (105) - Cudnn Error in initializeCommonContext: 4 (Could not initialize cudnn, please check cudnn installation.)
```

This usually happens when TRT can't run at all. The reason could vary from bad GPU or OOM.
It could be also because of having multiple GPUs (due to a known issue in TF-TRT). Could you rerun the test with 1 gpu?

**Comment by sanjoy** [comment]
> This usually happens when TRT can't run at all. The reason could vary from bad GPU or OOM.
> It could be also because of having multiple GPUs (due to a known issue in TF-TRT). Could you rerun the test with 1 gpu?

My machine does have two GPUs so that's probably it, but I haven't checked.

**Comment by sanjoy** [comment]
> My machine does have two GPUs so that's probably it, but I haven't checked.

Just to clarify, the next step here is: someone ( @DEKHTIARJonathan ?) needs to commandeer this PR and apply the [patch](https://github.com/tensorflow/tensorflow/pull/34919#issuecomment-580999679) from the previous comment.  Then the tests should start passing and we can continue reviewing the PR.

**Comment by double344931987** [comment]
I tried on single GPU, the convert step works very well but calibrate step costs all of the GPU memory (6GB) and shows the following error:

E tensorflow/compiler/tf2tensorrt/utils/trt_logger.cc:41] DefaultLogger ../rtSafe/safeContext.cpp (105) - Cudnn Error in initializeCommonContext: 4 (Could not initialize cudnn, please check cudnn installation.)

I think it is because tftrt use out of memory. So, is there some method to limit the memory cost during calibrate step.

**Comment by tfeher** [comment]
I have applied the patch from @sanjoy and rebased to latest master. The error in TRTEngineOpTestBase.AllowBuildAtRuntime was actially an error in defining the test: I have corrected the check for empty engine creation.  I have not encountered any errors during calibration. 

To fix CLA issues (original PR author has left the company), I have pushed the corrected branch to my repository, and opened a new PR #36852. This PR can be closed.

**Comment by gbaned** [comment]
Thanks @tfeher for the confirmation. 
Closing this PR, as its change has been moved to [#36852](https://github.com/tensorflow/tensorflow/pull/36852)

**Pull request closed without merging**